From 0287b7588b10d462205dc5045002e133b788f955 Mon Sep 17 00:00:00 2001
From: Myron Stowe <myron.stowe@redhat.com>
Date: Tue, 15 Jul 2014 03:52:53 -0400
Subject: [pci] pciehp: Clear Data Link Layer State Changed during init

Message-id: <20140715035253.4615.36858.stgit@amt.stowe>
Patchwork-id: 85804
O-Subject: [RHEL7.1 PATCH] PCI: pciehp: Clear Data Link Layer State Changed during init
Bugzilla: 1101097
RH-Acked-by: Dean Nelson <dnelson@redhat.com>
RH-Acked-by: Mark Langsdorf <mlangsdo@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1101097
Upstream Status: 0d25d35c987d7b0b63368d9c1ae35a917e1a7bab (pci/next)
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7700190
Testing: Testing by Dell was successful.

This patch is in Bjorn's 'pci/next' branch -
https://git.kernel.org/cgit/linux/kernel/git/helgaas/pci.git/log/?h=next -
and is targeted for being pulled upstream during the kernel's v3.17 merge
window.

commit 0d25d35c987d7b0b63368d9c1ae35a917e1a7bab
Author: Myron Stowe <myron.stowe@redhat.com>
Date:   Tue Jun 17 13:27:34 2014 -0600

    PCI: pciehp: Clear Data Link Layer State Changed during init

    During PCIe hot-plug initialization - pciehp_probe() - data structures
    related to slot capabilities are set up.  As part of this set up, ISRs are
    put in place to handle slot events and all event bits are cleared out.

    This patch adds the Data Link Layer State Changed (PCI_EXP_SLTSTA_DLLSC)
    Slot Status bit to the event bits that are cleared out during
    initialization.

    If the BIOS doesn't clear DLLSC before handoff to the OS, pciehp notices
    that it's set and interprets it as a new Link Up event, which results in
    spurious messages:

      pciehp 0000:82:04.0:pcie24: slot(4): Link Up event
      pciehp 0000:82:04.0:pcie24: Device 0000:83:00.0 already exists at 0000:83:00, cannot hot-add
      pciehp 0000:82:04.0:pcie24: Cannot add device at 0000:83:00

    Prior to e48f1b67f668 ("PCI: pciehp: Use link change notifications for
    hot-plug and removal"), pciehp ignored DLLSC.

    Reference:
      PCI-SIG.  PCI Express Base Specification Revision 4.0 Version 0.3
      (PCI-SIG, 2014): 7.8.11. Slot Status Register (Offset 1Ah).

    [bhelgaas: add e48f1b67f668 ref and stable tag]
    Fixes: e48f1b67f668 ("PCI: pciehp: Use link change notifications for hot-plug and removal")
    Link: https://bugzilla.kernel.org/show_bug.cgi?id=79611
    Signed-off-by: Myron Stowe <myron.stowe@redhat.com>
    Signed-off-by: Bjorn Helgaas <bhelgaas@google.com>
    CC: stable@vger.kernel.org  # v3.15+

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/pci/hotplug/pciehp_hpc.c b/drivers/pci/hotplug/pciehp_hpc.c
index 02b5b43..d3ee27a 100644
--- a/drivers/pci/hotplug/pciehp_hpc.c
+++ b/drivers/pci/hotplug/pciehp_hpc.c
@@ -795,7 +795,7 @@ struct controller *pcie_init(struct pcie_device *dev)
  pcie_capability_write_word(pdev, PCI_EXP_SLTSTA,
   PCI_EXP_SLTSTA_ABP | PCI_EXP_SLTSTA_PFD |
   PCI_EXP_SLTSTA_MRLSC | PCI_EXP_SLTSTA_PDC |
-  PCI_EXP_SLTSTA_CC);
+  PCI_EXP_SLTSTA_CC | PCI_EXP_SLTSTA_DLLSC);
 
  /* Disable software notification */
  pcie_disable_notification(ctrl);
-- 
1.7.1